File: csharp/collatz-conjecture/CollatzConjecture.cs
using System;

public static class CollatzConjecture
{
    public static int Steps(int number)
    {
        int n = number;
        int steps = 0;

        if (n <= 0)
        {
            throw new ArgumentOutOfRangeException("the number must be greater than zero");
        }

        while (n != 1)
        {
            if (n % 2 == 0)
            {
                n = n / 2;
            } else
            {
                n = (3 * n ) + 1;
            }
            steps++;
        }
        return steps;
    }
}

File: go/collatz-conjecture/collatz_conjecture.go
/*Package collatzconjecture handles the 3x+1 problem which can be summarized as follows:

Take any positive integer n. If n is even, divide n by 2 to get n / 2. If n is
odd, multiply n by 3 and add 1 to get 3n + 1. Repeat the process indefinitely.
The conjecture states that no matter which number you start with, you will
always reach 1 eventually.

*/
package collatzconjecture

import (
	"fmt"
)

// CollatzConjecture takes a number and runs it through the 3n+1 problem.
// Returns an error if the number is less than one.
func CollatzConjecture(number int) (int, error) {
	if number < 1 {
		return 0, fmt.Errorf("%d is not greater than one", number)
	}
	stepCount := 0

	for {
		if number == 1 {
			break
		}
		if number%2 == 0 {
			number /= 2
		} else {
			number = 3*number + 1
		}
		stepCount++
	}
	return stepCount, nil
}
